### DISCUSSION TITLE
particular constraints

### DISCUSSION THREAD

**Discussion by Elyos59** [dvdoug/BoxPacker#261] [Q&A]
Hi!

To begin with, I wanted to congratulate you for this very good work.

I have some packing restriction and I don't know how to implement it.

I have 4 kind of item :
1/ Not stackable & keep flat
2/ Not stackable & 3D rotation (useless for the moment)
3/ Stackable & keep flat (useless for the moment)
4/ Stackable & 3D rotation

For the moment, I focus on the 1st and the 4th kind of item.

The 1st can't have another item on it.
The 4th can have one or more items on it only if : 
- same dimensions or less (an item 2x2 can have 2 items 1x1 or 1 item 1x2 or 1 item 2x1 or 1 item 2x2 on it)
AND
- same weight or less (an item of 20kgs can have on it 3 items of 20 kgs or less each)

I think I have to use ConstrainedPlacementItem Class but I don't know how to cause of item below (dimensions/orientations/weight...)

Thanks for any help ;)

**Comment by dvdoug** [comment]
Hello @Elyos59 

https://boxpacker.io/en/stable/advanced-usage.html#example-don-t-allow-batteries-to-be-stacked would probably a good starting point - that example loops through the already packed items to find relevant items that would cause stacking issues. You'd need to modify that to add on your particular additional constraints

**Comment by Elyos59** [comment]
Ok.

I tried that : 
```
<?php
declare(strict_types=1);

namespace DVDoug\BoxPacker\Test;

use DVDoug\BoxPacker\Box;
use DVDoug\BoxPacker\ConstrainedPlacementItem;
use DVDoug\BoxPacker\PackedBox;
use DVDoug\BoxPacker\PackedItem;
use DVDoug\BoxPacker\Test\TestItem;

class NotStackableTestItem extends TestItem implements ConstrainedPlacementItem
{

	/**
	 *
	 * @param  Box            $box
	 * @param  PackedItemList $alreadyPackedItems
	 * @param  int            $proposedX
	 * @param  int            $proposedY
	 * @param  int            $proposedZ
	 * @param  int            $width
	 * @param  int            $length
	 * @param  int            $depth
	 * @return bool
	 */
	public function canBePacked(
        PackedBox $packedBox,
        int $proposedX,
        int $proposedY,
        int $proposedZ,
        int $width,
        int $length,
        int $depth
    ): bool {
        return ($proposedZ === 0);
	}
}
```
```
<?php
declare(strict_types=1);

namespace DVDoug\BoxPacker\Test;

use DVDoug\BoxPacker\Box;
use DVDoug\BoxPacker\ConstrainedPlacementItem;
use DVDoug\BoxPacker\PackedBox;
use DVDoug\BoxPacker\PackedItem;
use DVDoug\BoxPacker\Test\TestItem;

class StackableTestItem extends TestItem implements ConstrainedPlacementItem
{

	/**
	 *
	 * @param  Box            $box
	 * @param  PackedItemList $alreadyPackedItems
	 * @param  int            $proposedX
	 * @param  int            $proposedY
	 * @param  int            $proposedZ
	 * @param  int            $width
	 * @param  int            $length
	 * @param  int            $depth
	 * @return bool
	 */
	public function canBePacked(
        PackedBox $packedBox,
        int $proposedX,
        int $proposedY,
        int $proposedZ,
        int $width,
        int $length,
        int $depth
    ): bool {
        $alreadyPackedType = array_filter(
            iterator_to_array($packedBox->getItems(), false),
            function (PackedItem $item) {
                return $item->getItem() instanceof StackableTestItem;
            }
        );
		if(!count($alreadyPackedType)){
			return true;
		}
		/** @var PackedItem $alreadyPacked */
		foreach ($alreadyPackedType as $alreadyPacked) {
			if (
				$alreadyPacked->getZ() + $alreadyPacked->getDepth() === $proposedZ &&
				$proposedX >= $alreadyPacked->getX() && $proposedX <= ($alreadyPacked->getX() + $alreadyPacked->getWidth()) &&
				($proposedX + $this->getWidth()) <= ($alreadyPacked->getX() + $alreadyPacked->getWidth()) &&
				$proposedY >= $alreadyPacked->getY() && $proposedY <= ($alreadyPacked->getY() + $alreadyPacked->getLength()) &&
				($proposedY + $this->getLength()) <= ($alreadyPacked->getY() + $alreadyPacked->getLength()) &&
				$this->getWeight() <= $alreadyPacked->getItem()->getWeight()
				) {
				return true;
			}
		}
		return false;
	}
}
```
My box : 175 x 280 x 180 (width x length x depth)
My items : 
- 2 not stackable / keep flat : 100 x 120 x 20
- 6 stackable / best fit : 55 x 280 x 30
- 4 stackable / best fit : 60 x 80 x 90

And it doesn't work.
It gives me 8 boxes while it fits in one box... 

![Sans-titre-1](https://user-images.githubusercontent.com/16877840/137297827-3b40d079-9dc0-4ae3-ae98-4f7c61532ab4.jpg)

**Reply by Elyos59** [reply]
In my StackableTestItem Class I've changed : 

`if(!count($alreadyPackedType){`
by
`if($proposedZ === 0){`

and now items are in 3 boxes.

**Reply by Elyos59** [reply]
Same items should be stacked together... =/

**Reply by Elyos59** [reply]
I've tried to pre-sort items like you said here : https://github.com/dvdoug/BoxPacker/issues/218#issuecomment-793083932 and force NotStackableItem height to full box height but the result it's not better and NotStackableItem are not placed at first (even if they are first in my pre-sorted items array)

Result without my personnal classes : 

> These items fitted into 2 box(es)
> Box : width 175 length 280 depth 180
> Stackable Item was packed at co-ordinate (0, 0, 0) with width : 30, length : 280, depth : 55
> Stackable Item was packed at co-ordinate (30, 0, 0) with width : 30, length : 280, depth : 55
> Stackable Item was packed at co-ordinate (60, 0, 0) with width : 30, length : 280, depth : 55
> Stackable Item was packed at co-ordinate (90, 0, 0) with width : 30, length : 280, depth : 55
> Stackable Item was packed at co-ordinate (120, 0, 0) with width : 30, length : 280, depth : 55
> Stackable Item was packed at co-ordinate (0, 0, 55) with width : 30, length : 280, depth : 55
> Box : width 175 length 280 depth 180
> Not Stackable Item was packed at co-ordinate (0, 0, 0) with width : 120, length : 100, depth : 180
> Not Stackable Item was packed at co-ordinate (0, 100, 0) with width : 100, length : 120, depth : 180
> Stackable Item was packed at co-ordinate (100, 100, 0) with width : 60, length : 80, depth : 90
> Stackable Item was packed at co-ordinate (100, 100, 90) with width : 60, length : 80, depth : 90
> Stackable Item was packed at co-ordinate (0, 220, 0) with width : 80, length : 60, depth : 90
> Stackable Item was packed at co-ordinate (0, 220, 90) with width : 80, length : 60, depth : 90

**Comment by Elyos59** [comment]
I progress... a bit ...

```
<?php
declare(strict_types=1);

namespace DVDoug\BoxPacker\Test;

use DVDoug\BoxPacker\Box;
use DVDoug\BoxPacker\ConstrainedPlacementItem;
use DVDoug\BoxPacker\PackedBox;
use DVDoug\BoxPacker\PackedItem;
use DVDoug\BoxPacker\Test\TestItem;

class StackableTestItem extends TestItem implements ConstrainedPlacementItem
{

	/**
	 *
	 * @param  Box            $box
	 * @param  PackedItemList $alreadyPackedItems
	 * @param  int            $proposedX
	 * @param  int            $proposedY
	 * @param  int            $proposedZ
	 * @param  int            $width
	 * @param  int            $length
	 * @param  int            $depth
	 * @return bool
	 */
	public function canBePacked(
        PackedBox $packedBox,
        int $proposedX,
        int $proposedY,
        int $proposedZ,
        int $width,
        int $length,
        int $depth
    ): bool {
		$box = $packedBox->getBox();
		/* $boxWidth = $box->getInnerWidth();
		$boxLength = $box->getInnerLength(); */
		$boxDepth = $box->getInnerDepth();
		$weight = $this->getWeight();
        $stackablePackedType = array_filter(
            iterator_to_array($packedBox->getItems(), false),
            function (PackedItem $item) {
                return $item->getItem() instanceof StackableTestItem;
            }
        );
		if(count($stackablePackedType) == 0 && $proposedZ === 0){
			return true;
		}
		$stackablePositions = array();
		/** @var PackedItem $alreadyPacked */
		foreach ($stackablePackedType as $stackablePacked) {
			list($x_,$y_,$z_,$width_,$length_,$depth_,$weight_) = array($stackablePacked->getX(),$stackablePacked->getY(),$stackablePacked->getZ(),$stackablePacked->getWidth(),$stackablePacked->getLength(),$stackablePacked->getDepth(),$stackablePacked->getItem()->getWeight());
			if($width_>=$width && $length_>=$length && $weight_>=$weight && ($z_+$depth_+$depth)<=$boxDepth){
				foreach($stackablePositions as $s => $stackablePosition){
					if($stackablePosition[$s]["x"]==$x_ && $stackablePosition[$s]["y"]==$y_){
						if($stackablePosition[$s]["z"]>$z_+$depth_){
							$stackablePosition[$s]["z"]=$z_+$depth_;
						}
						continue;
					}
				}
				$stackablePositions[] = array("x"=>$x_,"y"=>$y_,"z"=>$z_+$depth_,"item"=>$stackablePacked);
			}
		}
		if(count($stackablePositions)){
			foreach($stackablePositions as $stackablePosition){
				if(
					$stackablePosition["z"] === $proposedZ &&
					$proposedX >= $stackablePosition["x"] && $proposedX <= ($stackablePosition["x"] + $stackablePosition["item"]->getWidth()) &&
					($proposedX + $width) <= ($stackablePosition["x"] + $stackablePosition["item"]->getWidth()) &&
					$proposedY >= $stackablePosition["y"] && $proposedY <= ($stackablePosition["y"] + $stackablePosition["item"]->getLength()) &&
					($proposedY + $length) <= ($stackablePosition["y"] + $stackablePosition["item"]->getLength())
				){
					return true;
				}
			}
		}else if($proposedZ === 0){
			return true;
		}
		return false;
	}
}
```
Always two boxes but the second one could fill the first. There is just one unfilled hole :

![Capture d’écran (23)](https://user-images.githubusercontent.com/16877840/137712430-6d5ee26f-7ccb-4a31-be80-5d5479820865.png)

**Comment by Elyos59** [comment]
I'm tearing my hair out!

No way to understand why it doesn't work ?!
